`azurerm_managed_application` - add support for managed identity

hqhqhqhqhqhqhqhqhqhqhq opened the pull request:
<!--  All Submissions -->

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR adds support for managed identity to the `azurerm_managed_application` resource. The identity functionality was already implemented in the codebase but was missing from the resource schema and documentation. This change enables users to configure system-assigned, user-assigned, or both types of managed identities for their managed applications.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass.

<img width="923" height="629" alt="image" src="https://github.com/user-attachments/assets/9f2b5796-e538-4c75-a6fb-a77ffd39c9a1" />
https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_MANAGEDAPPLICATIONS/682763?buildTab=tests



## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_managed_application` - support for the `identity` property

## Related Issue(s)

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

This PR adds managed identity support to managed applications, which enhances security by enabling:
- System-assigned identities for automatic credential management
- User-assigned identities for shared identity scenarios
- Both identity types for comprehensive access control

The identity functionality follows the same patterns used by other Azure resources in the provider and maintains consistency with Azure's managed identity best practices.

<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

hqhqhqhqhqhqhqhqhqhqhq committed:
Update managed_application_resource.go

hqhqhqhqhqhqhqhqhqhqhq committed:
implent code, test, doc

hqhqhqhqhqhqhqhqhqhqhq committed:
Refactor identity handling and update related tests

Simplifies the handling of the identity type in flattenManagedApplicationIdentity by removing unnecessary string conversion. Updates acceptance tests to remove checks for specific identity attributes, focusing instead on resource existence in Azure.

hqhqhqhqhqhqhqhqhqhqhq committed:
Update managed_application_resource_test.go

ms-henglu commented on internal/services/managedapplications/managed_application_resource_test.go:
```diff
@@ -153,6 +153,58 @@ func TestAccManagedApplication_tags(t *testing.T) {
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
```
Can you add more steps to test it can change from system assigned identity to user assigned identity, then change back to none identity?

ms-henglu reviewed (changes requested):
Thanks for the PR! LGTM, just one minor suggestion.

hqhqhqhqhqhqhqhqhqhqhq committed:
Update managed_application_resource_test.go

ms-henglu reviewed (approved):
Thanks for the PR! LGTM.

liuwuliuyun commented on internal/services/managedapplications/managed_application_resource.go:
```diff
@@ -462,3 +487,67 @@ func compactParameterOrOutputValue(v interface{}) (string, error) {
+		for k, v := range expanded.IdentityIds {
+			userAssignedIdentities[k] = applications.UserAssignedResourceIdentity{
+				PrincipalId: v.PrincipalId,
+				TenantId:    v.ClientId,
```
Hi, just curious—why is it assigned as v.clientid instead of v.tenentid?

hqhqhqhqhqhqhqhqhqhqhq replied:
Good catch - this was a bug. The API's UserAssignedResourceIdentity only has PrincipalId and TenantId fields (no ClientId). I've fixed the mapping to only use PrincipalId since ClientId isn't exposed by the managed applications API.

liuwuliuyun commented on internal/services/managedapplications/managed_application_resource.go:
```diff
@@ -462,3 +487,67 @@ func compactParameterOrOutputValue(v interface{}) (string, error) {
+		if input.UserAssignedIdentities != nil {
+			for k, v := range *input.UserAssignedIdentities {
+				identityIds[k] = identity.UserAssignedIdentityDetails{
+					ClientId:    v.TenantId,
```
Same here

liuwuliuyun commented on website/docs/r/managed_application.html.markdown:
```diff
@@ -103,6 +105,18 @@ The `plan` block exports the following:
+
+* `identity_ids` - (Optional) Specifies a list of User Assigned Managed Identity IDs to be assigned to this Managed Application.
+
+~> **Note:** When `type` is set to `SystemAssigned`, The assigned `principal_id` and `tenant_id` can be retrieved after the Managed Application has been created. More details are available below.
```
It mentions that more details are available below, but I can't seem to find them in this PR.

liuwuliuyun reviewed (commented):
Just a few tweaks or a quick explanation, and we're all set.

hqhqhqhqhqhqhqhqhqhqhq committed:
Address liuwuliuyun review comments for managed_application identity

- Fix ClientId/TenantId field mapping bug in expand function
  The API UserAssignedResourceIdentity only has PrincipalId and TenantId (no ClientId)
  Removed incorrect ClientId mapping that was causing compile errors

- Fix flatten function to only map PrincipalId from API response
  The managed applications API doesn't expose ClientId for user-assigned identities

- Add Attributes Reference section to documentation
  Documents principal_id and tenant_id as exported attributes
  Clarifies when these values are available for SystemAssigned identity

- Remove duplicate note from Arguments section
  Consolidated identity attribute documentation in Attributes Reference section

This fixes the incorrect field mappings that would have caused runtime errors.

hqhqhqhqhqhqhqhqhqhqhq committed:
Update managed_application_resource.go

liuwuliuyun reviewed (approved):
LGTM!

sreallymatt reviewed (changes requested):
Thanks @hqhqhqhqhqhqhqhqhqhqhq - just a couple comments, the added [test is also failing](https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_MANAGEDAPPLICATIONS/639810?buildTab=tests)

hqhqhqhqhqhqhqhqhqhqhq committed:
azurerm_managed_application - address identity review feedback

hqhqhqhqhqhqhqhqhqhqhq committed:
azurerm_managed_application - handle identity removal

hqhqhqhqhqhqhqhqhqhqhq commented on internal/services/managedapplications/managed_application_resource.go:
```diff
@@ -462,3 +487,63 @@ func compactParameterOrOutputValue(v interface{}) (string, error) {
+
+	if input != nil {
+		config = &identity.SystemAndUserAssignedMap{
+			Type:        identity.Type(*input.Type),
```
thanks, `Type` is now nil-checked

hqhqhqhqhqhqhqhqhqhqhq commented on website/docs/r/managed_application.html.markdown:
```diff
@@ -103,6 +105,16 @@ The `plan` block exports the following:
+
+* `identity_ids` - (Optional) Specifies a list of User Assigned Managed Identity IDs to be assigned to this Managed Application.
+
+~> **Note:** The `identity_ids` is required when `type` is set to `UserAssigned` or `SystemAssigned, UserAssigned`.
```
thx, updated
